File: web/tests/clients-view.test.js
/**
 * Headless tests for web/js/lib/clients-view.js (WP 4a.7 DoD).
 *
 * Run: node --test "web/tests/*.test.js"   (see web/tests/README.md)
 */
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  partitionClients,
  hitRatePercent,
  addressesText,
  describeHealthyClient,
  describeBypassClient,
  bypassBannerText,
} from "../js/lib/clients-view.js";

function client(overrides) {
  return {
    client_id: "workshop-pc",
    first_seen: "2026-08-01T00:00:00Z",
    last_reported_at: "2026-08-10T00:00:00Z",
    app_count: 10,
    source_addrs: ["10.10.0.21"],
    cache_hits: 5,
    cache_misses: 1,
    bytes_served: 1234,
    last_seen_in_cache_log: "2026-08-10T00:00:00Z",
    bypass_suspected: false,
    ...overrides,
  };
}

// ---------------------------------------------------------------------
// partitionClients
// ---------------------------------------------------------------------

test("partitionClients: splits by bypass_suspected, preserving order within each bucket", () => {
  const a = client({ client_id: "a", bypass_suspected: true });
  const b = client({ client_id: "b", bypass_suspected: false });
  const c = client({ client_id: "c", bypass_suspected: true });
  const { bypassing, healthy } = partitionClients([a, b, c]);
  assert.deepEqual(bypassing.map((x) => x.client_id), ["a", "c"]);
  assert.deepEqual(healthy.map((x) => x.client_id), ["b"]);
});

test("partitionClients: empty/null/undefined input -> both buckets empty", () => {
  for (const input of [[], null, undefined]) {
    const { bypassing, healthy } = partitionClients(input);
    assert.deepEqual(bypassing, []);
    assert.deepEqual(healthy, []);
  }
});

// ---------------------------------------------------------------------
// hitRatePercent
// ---------------------------------------------------------------------

test("hitRatePercent: rounds hits/(hits+misses) to a whole-number percentage", () => {
  assert.equal(hitRatePercent(client({ cache_hits: 96, cache_misses: 4 })), 96);
  assert.equal(hitRatePercent(client({ cache_hits: 1, cache_misses: 2 })), 33); // 33.33.. -> 33
  assert.equal(hitRatePercent(client({ cache_hits: 2, cache_misses: 1 })), 67); // 66.67 -> 67
});

test("hitRatePercent: null (never fabricated 0%) when there are zero total requests", () => {
  assert.equal(hitRatePercent(client({ cache_hits: 0, cache_misses: 0 })), null);
});

test("hitRatePercent: non-finite/negative/missing counters treated as 0, not NaN/crash", () => {
  // NaN/negative hits count as 0 -> 0 hits out of 5 total (the misses) = 0%.
  assert.equal(hitRatePercent(client({ cache_hits: NaN, cache_misses: 5 })), 0);
  assert.equal(hitRatePercent(client({ cache_hits: -3, cache_misses: 5 })), 0);
  assert.equal(hitRatePercent({}), null);
  assert.equal(hitRatePercent(null), null);
});

// ---------------------------------------------------------------------
// addressesText
// ---------------------------------------------------------------------

test("addressesText: joins multiple addresses", () => {
  assert.equal(
    addressesText(client({ source_addrs: ["10.10.0.21", "10.10.0.22"] })),
    "10.10.0.21, 10.10.0.22",
  );
});

test("addressesText: honest 'no known address' for an empty/missing list (pre-schema-v9 reports)", () => {
  assert.equal(addressesText(client({ source_addrs: [] })), "no known address");
  assert.equal(addressesText(client({ source_addrs: undefined })), "no known address");
  assert.equal(addressesText(null), "no known address");
});

// ---------------------------------------------------------------------
// describeHealthyClient / describeBypassClient
// ---------------------------------------------------------------------

test("describeHealthyClient: full stats line", () => {
  const line = describeHealthyClient(
    client({ app_count: 61, bytes_served: 41.8 * 1_073_741_824, cache_hits: 96, cache_misses: 4 }),
  );
  assert.equal(line, "61 games reported · 41.8 GB served · 96% hit");
});

test("describeHealthyClient: singular 'game' for app_count === 1", () => {
  const line = describeHealthyClient(client({ app_count: 1, cache_hits: 1, cache_misses: 0 }));
  assert.match(line, /^1 game reported/);
});

test("describeHealthyClient: honest fallbacks when nothing has happened yet", () => {
  const line = describeHealthyClient(
    client({ app_count: null, bytes_served: 0, cache_hits: 0, cache_misses: 0 }),
  );
  assert.equal(line, "game count unknown · nothing served yet · no cache requests yet");
});

test("describeBypassClient: states the observation, not a verdict", () => {
  const line = describeBypassClient(client({ app_count: 38 }));
  assert.equal(line, "38 games reported · none of its downloads have reached the cache recently");
  assert.doesNotMatch(line, /DNS/i); // the cause-agnostic explanation lives in BYPASS_EXPLANATION, not here
});

// ---------------------------------------------------------------------
// bypassBannerText
// ---------------------------------------------------------------------

test("bypassBannerText: empty string when nobody is suspected", () => {
  assert.equal(bypassBannerText([client(), client({ client_id: "b" })]), "");
  assert.equal(bypassBannerText([]), "");
  assert.equal(bypassBannerText(null), "");
});

test("bypassBannerText: singular wording, names the one client", () => {
  assert.equal(
    bypassBannerText([client({ client_id: "DEMON", bypass_suspected: true })]),
    "Client DEMON is bypassing the cache — check its DNS.",
  );
});

test("bypassBannerText: plural wording for two or more, does not name them", () => {
  const clients = [
    client({ client_id: "a", bypass_suspected: true }),
    client({ client_id: "b", bypass_suspected: true }),
  ];
  assert.equal(bypassBannerText(clients), "2 clients are bypassing the cache — check their DNS.");
});

File: app/app/src/test/java/dev/steamvault/app/ui/clients/logic/ClientsViewTest.kt
package dev.steamvault.app.ui.clients.logic

import dev.steamvault.app.net.model.ClientOut
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotSame
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Ports `web/tests/clients-view.test.js`'s `partitionClients`/
 * `hitRatePercent` cases onto [ClientOut], plus the WP 4b.10 brief's
 * [ClientRowModel] stability claim -- see that class's kdoc for why this
 * is the Compose-platform equivalent of `clients-render-plan.js`'s full/
 * patch/rebuild verdict, same class of proof
 * `ui/downloads/logic/JobCardModelTest.kt` establishes for [JobCardModel]-
 * shaped stability guarantees.
 */
class ClientsViewTest {

    private fun client(
        clientId: String = "workshop-pc",
        appCount: Int? = 10,
        sourceAddrs: List<String> = listOf("10.10.0.21"),
        cacheHits: Int = 5,
        cacheMisses: Int = 1,
        bytesServed: Long = 1234L,
        bypassSuspected: Boolean = false,
    ) = ClientOut(
        client_id = clientId,
        first_seen = "2026-08-01T00:00:00Z",
        last_reported_at = "2026-08-10T00:00:00Z",
        app_count = appCount,
        source_addrs = sourceAddrs,
        cache_hits = cacheHits,
        cache_misses = cacheMisses,
        bytes_served = bytesServed,
        last_seen_in_cache_log = "2026-08-10T00:00:00Z",
        bypass_suspected = bypassSuspected,
    )

    // -------------------------------------------------------------
    // partitionClients
    // -------------------------------------------------------------

    @Test
    fun `partitionClients splits by bypass_suspected, preserving order within each bucket`() {
        val a = client(clientId = "a", bypassSuspected = true)
        val b = client(clientId = "b", bypassSuspected = false)
        val c = client(clientId = "c", bypassSuspected = true)
        val partition = partitionClients(listOf(a, b, c))
        assertEquals(listOf("a", "c"), partition.bypassing.map { it.client_id })
        assertEquals(listOf("b"), partition.healthy.map { it.client_id })
    }

    @Test
    fun `partitionClients on an empty list produces two empty buckets`() {
        val partition = partitionClients(emptyList())
        assertTrue(partition.bypassing.isEmpty())
        assertTrue(partition.healthy.isEmpty())
    }

    // -------------------------------------------------------------
    // hitRatePercent
    // -------------------------------------------------------------

    @Test
    fun `hitRatePercent rounds hits over hits-plus-misses to a whole-number percentage`() {
        assertEquals(96, hitRatePercent(client(cacheHits = 96, cacheMisses = 4)))
        assertEquals(33, hitRatePercent(client(cacheHits = 1, cacheMisses = 2))) // 33.33.. -> 33
        assertEquals(67, hitRatePercent(client(cacheHits = 2, cacheMisses = 1))) // 66.67 -> 67
    }

    @Test
    fun `hitRatePercent is null, never a fabricated 0 percent, for zero total requests`() {
        assertNull(hitRatePercent(client(cacheHits = 0, cacheMisses = 0)))
    }

    @Test
    fun `hitRatePercent treats a negative counter as 0, not a negative or crashing rate`() {
        // -3 hits / 5 misses -> 0 hits out of a 5 total -> 0%, mirrors
        // clients-view.js's safeCount guard (Kotlin's non-nullable Int has
        // no NaN case to mirror).
        assertEquals(0, hitRatePercent(client(cacheHits = -3, cacheMisses = 5)))
    }

    // -------------------------------------------------------------
    // ClientRowModel stability (WP 4b.10 brief)
    // -------------------------------------------------------------

    @Test
    fun `two ticks of a genuinely-unchanged client produce an EQUAL model from DISTINCT instances`() {
        val tick1 = buildClientRowModel(client())
        val tick2 = buildClientRowModel(client())
        assertNotSame(tick1, tick2)
        assertEquals(tick1, tick2)
    }

    @Test
    fun `a stats-only diff changes ONLY the stats field`() {
        val before = buildClientRowModel(client(cacheHits = 5, cacheMisses = 1, bytesServed = 1234L))
        val after = buildClientRowModel(client(cacheHits = 96, cacheMisses = 4, bytesServed = 999_999L))

        assertTrue(before != after) // MUTATION TARGET: a real diff must be observed at all
        assertEquals(before.copy(stats = after.stats), after) // ... but nowhere else.
        assertEquals(before.clientId, after.clientId)
        assertEquals(before.bypassSuspected, after.bypassSuspected)
        assertEquals(before.addresses, after.addresses)
    }

    @Test
    fun `a bypass_suspected flip changes ONLY that field, never clientId or addresses`() {
        val before = buildClientRowModel(client(bypassSuspected = false))
        val after = buildClientRowModel(client(bypassSuspected = true))

        assertTrue(before != after) // MUTATION TARGET
        assertEquals(before.copy(bypassSuspected = after.bypassSuspected), after)
        assertEquals(before.clientId, after.clientId)
        assertEquals(before.addresses, after.addresses)
    }

    @Test
    fun `buildClientRowModel carries app_count, bytes_served and the computed hit rate into stats`() {
        val model = buildClientRowModel(client(appCount = 61, cacheHits = 96, cacheMisses = 4, bytesServed = 44_000_000_000L))
        assertEquals(61, model.stats.gamesReported)
        assertEquals(44_000_000_000L, model.stats.bytesServed)
        assertEquals(96, model.stats.hitRatePercent)
    }

    @Test
    fun `buildClientRowModel reports app_count null as null, never a fabricated zero`() {
        val model = buildClientRowModel(client(appCount = null))
        assertNull(model.stats.gamesReported)
    }
}
